File: Problems/Week_1/3.24_괄호(9012)/임용식/9012.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AlgoCSharp
{
    class P2885
    {
        static void Main(string[] args)
        {
            int t;
            StringBuilder sb = new StringBuilder();
            t = Convert.ToInt32(Console.ReadLine());

            while (t-- > 0)
            {
                int stack = 0;
                bool isValid = true;
                string ps = Console.ReadLine();

                foreach (char p in ps)
                {
                    if (p == '(')
                    {
                        stack++;
                    }
                    else if (stack > 0)
                    {
                        stack--;
                    }
                    else
                    {
                        isValid = false;
                        break;
                    }
                }
                if (stack > 0)
                {
                    isValid = false;
                }
                sb.Append((isValid ? "YES\n" : "NO\n"));
            }
            Console.WriteLine(sb);
        }
    }
}

File: Problems/Week_1/3.24_괄호(9012)/임용식/9012.cpp
#include <iostream>
#include <string>

int main()
{
	int t;
	std::cin >> t;

	while (t--)
	{
		int stack = 0;
		bool isValid = true;
		std::string ps;
		std::cin >> ps;

		for (const auto& p : ps)
		{
			if (p == '(')
			{
				stack++;
			}
			else if (stack != 0)
			{
				stack--;
			}
			else
			{
				isValid = false;
				break;
			}
		}
		if (stack != 0)
		{
			isValid = false;
		}
		std::cout << (isValid ? "YES" : "NO")<<'\n';
	}

	return 0;
}